src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,347 @@
+                }
+
+                return new PooledConnection(this, serviceName, (JsonRpcConnection)await TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken).ConfigureAwait(false));
+            }
+
+            private async Task<Connection> TryCreateNewConnectionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
+            {
+                var dataRpc = _remotableDataRpc.TryAddReference();
+                if (dataRpc == null)
+                {
+                    // dataRpc is disposed. this can happen if someone killed remote host process while there is
+                    // no other one holding the data connection.
+                    // in those error case, don't crash but return null. this method is TryCreate since caller expects it to return null
+                    // on such error situation.
+                    return null;
+                }
+
+                // get stream from service hub to communicate service specific information
+                // this is what consumer actually use to communicate information
+                var serviceStream = await Connections.RequestServiceAsync(_hubClient, serviceName, _hostGroup, _timeout, cancellationToken).ConfigureAwait(false);
+
+                return new JsonRpcConnection(_hubClient.Logger, callbackTarget, serviceStream, dataRpc);
+            }
+
+            private void Free(string serviceName, JsonRpcConnection connection)
+            {
+                using (_shutdownLock.DisposableRead())
+                {
+                    if (!_usePool)
+                    {
+                        // pool is not being used.
+                        connection.Dispose();
+                        return;
+                    }
+
+                    // queue must exist
+                    var queue = _pools[serviceName];
+                    if (queue.Count >= _maxPoolConnections)
+                    {
+                        // let the connection actually go away
+                        connection.Dispose();
+                        return;
+                    }
+
+                    // pool the connection
+                    queue.Enqueue(connection);
+                }
+            }
+
+            public void Shutdown()
+            {
+                using (_shutdownLock.DisposableWrite())
+                {
+                    // mark not to use pool
+                    _usePool = false;
+
+                    // let ref count this one is holding go
+                    _remotableDataRpc.Dispose();
+
+                    // let all connections in the pool to go away
+                    foreach (var kv in _pools)
+                    {
+                        while (kv.Value.TryDequeue(out var connection))
+                        {
+                            connection.Dispose();
+                        }
+                    }
+                }
```

Comment:
clear the queues, and then clear the entire dictionary.

Reply:
it will naturally go away. when RemoteHostClient goes away.

Reply:
> it will naturally go away. when RemoteHostClient goes away.

I really don't like that.  You disposed resources.  Keeping htem around in places where someone might actually stumble over them is dangerous.  The code may be ok now, but in the future if this class changes, it's a potential footgun.

Reply:
oh well. sure. even though not sure why fixing this one place make anything more robust when I see these thing all over the roslyn..

Reply:
> oh well. sure. even though not sure why fixing this one place make anything more robust 

Well, this code becomes more robust.  So by definition things are now more robust :)

> I see these thing all over the roslyn..

Don't care :)   If this is all over the place, we should fix those as well.  However, here's the important piece for me:  

This OOP/ServiceHub work is literally critical infrastructure of the Roslyn Workspace. It has a higher bar than much of the rest of the Roslyn.  We should be putting in the extra care and effort to make it as rock solid as possible, and we shouldn't use other, less well written, parts of roslyn as a reason to not do that here.